test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeAvailabilityTest.java

```diff
@@ -0,0 +1,137 @@
+                         new AvailabiltyTester(ALL, ONE));
+    }
+
+    public void testAvailability(AvailabiltyTester... testers) throws Throwable
+    {
+        new TestCase()
+        .nodes(3)
+        .nodesToUpgrade(1)
+        .upgrade(Versions.Major.v22, Versions.Major.v30)
+        .upgrade(Versions.Major.v3X, Versions.Major.v4)
+        .upgrade(Versions.Major.v30, Versions.Major.v4)
+        .withConfig(config -> config.set("read_request_timeout_in_ms", SECONDS.toMillis(30))
+                                    .set("write_request_timeout_in_ms", SECONDS.toMillis(30)))
+        .setup(c -> c.schemaChange(withKeyspace("CREATE TABLE %s.tbl (k uuid, c int, v int, PRIMARY KEY (k, c))")))
+        .runAfterNodeUpgrade((cluster, n) -> {
+
+            // using an upgraded and a not upgraded coordinator...
+            for (int node = 1; node < cluster.size(); node++)
+            {
+                ICoordinator coordinator = cluster.coordinator(node);
+                int nodeDown = node;
+
+                // using 0 to 2 down nodes...
+                for (int numNodesDown = 0; numNodesDown < cluster.size(); numNodesDown++)
+                {
+                    // disable communications to the down nodes
+                    if (numNodesDown > 0)
+                    {
+                        nodeDown = nextNode(nodeDown, cluster.size());
+                        cluster.filters().verbs(READ_REQ.id, MUTATION_REQ.id).to(nodeDown).drop();
+                    }
+
+                    // run the test cases that are compatible with the number of down nodes
+                    for (AvailabiltyTester tester : testers)
+                    {
+                        tester.test(coordinator, numNodesDown);
+                    }
+                }
+
+                // restore communication to all nodes
+                cluster.filters().reset();
+            }
+        }).run();
+    }
+
+    private static int nextNode(int node, int numNodes)
+    {
+        return node == numNodes ? 1 : node + 1;
+    }
+
+    private static class AvailabiltyTester
+    {
+        private static final String INSERT = withKeyspace("INSERT INTO %s.tbl (k, c, v) VALUES (?, ?, ?)");
+        private static final String SELECT = withKeyspace("SELECT * FROM %s.tbl WHERE k = ?");
+
+        private final ConsistencyLevel writeConsistencyLevel;
+        private final ConsistencyLevel readConsistencyLevel;
+
+        private AvailabiltyTester(ConsistencyLevel writeConsistencyLevel, ConsistencyLevel readConsistencyLevel)
+        {
+            this.writeConsistencyLevel = writeConsistencyLevel;
+            this.readConsistencyLevel = readConsistencyLevel;
+        }
+
+        public void test(ICoordinator coordinator, int numNodesDown)
+        {
+            if (numNodesDown > maxNodesDown(writeConsistencyLevel) ||
+                numNodesDown > maxNodesDown(readConsistencyLevel))
+                return;
```

Comment:
@adelapena What do you think about testing the negative cases here as well (i.e. just catching the errors and verifying them)? I don't imagine the verb filtering will suddenly break, but you know...paranoia...

Reply:
@maedhroz I initially I added negative cases, but waiting for the read/write timeouts seemed to make the tests a bit unstable, at least locally. I think that for upgrade tests I'd prefer to live without negative tests and with a limited number of consistency levels, but I'm open to give it another go.

Reply:
That would be a long wait, especially given we've increased the timeouts above. Would simply shutting down the affected nodes result in quicker failures?

Reply:
Shutting the nodes down might solve the problem of waiting for the timeouts. However I'm finding problems restarting multiple nodes, for example:
```java
new TestCase()
.nodes(3)
.nodesToUpgrade(1)
.upgrade(Versions.Major.v3X, Versions.Major.v4)
.setup(c -> {})
.runAfterNodeUpgrade((cluster, node) -> {
    cluster.get(1).shutdown().get();
    cluster.get(2).shutdown().get();
    cluster.get(1).startup();
    cluster.get(2).startup();
}).run();
```
That produces an `IllegalStateException: Can't use shut down instances, delegate is null` error, when I understand that it should work, unless I'm missing something.

Reply:
A workaround for shutting down the nodes is, given that it seems we can't bring them up again, is to run two separate upgrade tests. One of them uses the upgraded node as coordinator and progressively shuts down the other two nodes, and the other uses one of the not upgraded nodes as coordinator and progressively shuts down the other two nodes. Done [here](https://github.com/apache/cassandra/pull/801/commits/e9d9fb5c91abe8a49ac10dba8b196d889dff3db6).